### PR TITLE
Move layout from Type_abstract to type_declaration

### PR DESCRIPTION
**Pull request by goldfirere**
This was extracted from #1340, but it's quite separate and makes more sense as its own PR.

This takes the layout of a type out from the `Type_abstract` node and puts it right in the `type_declaration`. I think it's a little cleaner this way, but the real motivation for this is to fix performance problems that arose in #1340.

### CODE REVIEW

**Comment by goldfirere** [comment]
Don't review yet. I have some improvements in the works.

**Comment by goldfirere** [comment]
The "improvements" might be controversial (though I like them), so I pushed them off to #1386. This is ready for review.

**Comment by goldfirere** [comment]
@ccasin observed that the layout in `Type_abstract` is very like the layout in `Record_unboxed` and `Variant_unboxed`, so I removed those, too, in the commits above.

**Review by ccasin — CHANGES_REQUESTED** [review]
I agree this is cleaner - I've left some comments and questions throughout.

However, I'm wondering about this program:

```ocaml
type t_void [@@void]

type ('a : any) t = A of 'a

type s = t_void t [@@immediate]
```

Obviously, we don't allow the `any` parameter today, but we plan to.  `s` should check, because `'a t` is immediate just if `'a` is void.

I don't actually think either the old code or the new code gets this quite right (even if we imagine they allowed the `('a : any)`).  The new approach is wrong because we just trust the stored layout on the decl of `t`.  The old approach was wrong because `layout_bound_of_kind` trusted the stored layout in the variant_representation - it didn't consider parameter instantiations.

One can imagine a few approaches to this. We could soup up `layout_bound_of_kind` to really recompute the thing but with parameters instantiated.  We could record dependencies on parameter layouts, either in the variant and record representations or in the overall decl (or both).

Maybe it would be more productive to talk through this example, but for the moment I'm less convinced about dropping `layout_bound_of_kind` (whether or not we move the layout for abstract types from the kind to the decl).

**Comment by goldfirere** [comment]
@ccasin:

> However, I'm wondering about this program:
> 
> ...
> 
> Maybe it would be more productive to talk through this example, but for the moment I'm less convinced about dropping `layout_bound_of_kind` (whether or not we move the layout for abstract types from the kind to the decl).

Yes, I've been worried about such programs, too. My very-sketchy plan was to actually make `type_layout : layout list -> layout` in the glorious future, where the input list comprises the layouts of the type arguments. That's not quite right, because most decl layouts are constant, so we'd ideally want `layout Lazy.t list -> layout` but more realistically probably some other more inspectable structure that says which argument layouts it really needs. Maybe the lazy one would work? Not sure how expensive making thunks is. I think this general plan would work, and would pay other dividends, too, in that (I think) we'd never have to expand in order to get a more precise layout (but I've gotten that interaction wrong before).

As you say, I think the problem you highlight here is orthogonal to this patch (but definitely a problem).

I don't think we can keep `layout_bound_of_kind` and also move the layout out from `Type_abstract`. Essentially, the action of `layout_bound_of_kind` would be closed over in my `layout Lazy.t list -> layout`.

File: ocaml/typing/typedecl.ml
Comment (goldfirere) [reply ocaml/typing/typedecl.ml:583]: Yes, that's a good point. See my expanded Note.

File: ocaml/typing/typedecl.ml
Comment (goldfirere) [reply ocaml/typing/typedecl.ml:1070]: Yes good point.

**Comment by goldfirere** [comment]
OK. I've pushed some updates, including a big new comment in Typedecl. That comment ends describing an infelicity of the current design -- but I claim that the infelicity predates this patch.

**Comment by ccasin** [comment]
On reflection, I don't feel super strongly about keeping `layout_bound_of_kind`, if we're going to put a layout in type declarations.  I think what I said above is true, but keeping the dead code because we're going to want it in the future is not the right solution.  We can evaluate whether we want to add back something like it when we deal with those complicated cases.  Maybe the comments explaining that the `type_layout` in decls is exact in the case of record and variant kinds should be updated to indicate this may not always be true.

Having thought about these future cases where we're going to need to deal with substitutions, I appreciate Stephen's original design more.  Having the layouts in the places where they matter and only there is parsimonious, and has minimal performance impact.  In a vacuum, I'd keep the way it's done now, but I don't feel strongly about it and you clearly do, so I think this PR is fine (I'll approve after I look through your responses to my comments).

This PR will make some of the info in record and variant representations dead code, as your subsequent PR shows.  But it's clear after having thought through these examples that we can't get rid of that info permanently.  And in fact I'm pretty sure we're going to want them sooner than that - even just for unboxed floats (at least, when we allow unboxed floats in structures).  So we should put that subsequent PR on hold - we can reevaluate after shipping the unboxed float stuff, in case I'm wrong about wanting this info there.

**Comment by goldfirere** [comment]
We seem to be converging here. As I said above, I'm not specifically against keeping `layout_bound_of_kind`, though -- as you say -- it would be dead code for a little while. And so I advocate removing it -- it's not very complicated. My suspicion is that when we get to the point of restoring the functionality in some way, things will look quite different than they do today, and we'll just solve the problem before us. One step at a time!

Part of that suspicion is informed by my suspicion that we can get rid of the need for expansion in layout calculations altogether, with the `layout Lazy.t list -> layout` approach. But I haven't worked that out and don't need to today. All in good time.

File: ocaml/typing/typedecl.ml
Comment (ccasin) [inline ocaml/typing/typedecl.ml:584]: I'm not sure I'd call this a flaw.  Examples like this are why we only promise that the stored layout will be an upper bound, and as you note it's not a problem because the appropriate expansions happen later.  But it's a nice illustration of why we can't promise something stronger - maybe revise the comment slightly to indicate that?

File: ocaml/typing/typedecl.ml
Comment (ccasin) [inline ocaml/typing/typedecl.ml:542]: will be updated in [update_decl_layout] or [check_coherence]

File: ocaml/typing/typedecl.ml
Comment (ccasin) [inline ocaml/typing/typedecl.ml:546]: Note that it's updated later in `check_coherence`?

(There's an additional subtlety this comment doesn't get into which is that any of the layout calculations mentioned may be inaccurate because the environment used in transl declaration doesn't have manifests for mutually defined types.  That's true about case 2 here, but is also true about the other cases.  It's right not to try to explain both things at once, but maybe include a reference to the comment in `enter_type` somewhere in here).

**Review by ccasin — APPROVED** [review]
OK.  I think we're agreed.  I have some suggestions about your nice new comment, but this can be merged once you're happy with that comment.

File: ocaml/typing/datarepr.ml
Comment (ccasin) [reply ocaml/typing/datarepr.ml:82]: Yes, my mistake.  I though this was a change in behavior, but I see the check I was thinking about is in `Typedecl.update_label_layouts`

**Review by gretay-js — APPROVED** [review]
for files owned by me.

**Pull request merged**
